### PR TITLE
Remove worker reconnect

### LINKED ISSUE #6350: Remove worker reconnection logic
It's proving difficult to merge a fix for https://github.com/dask/distributed/issues/5480. https://github.com/dask/distributed/pull/6329, https://github.com/dask/distributed/pull/6272, and https://github.com/dask/distributed/pull/6341 are causing tests to fail in unexpected and hard-to-debug ways. Because reconnection is quite broken as things stand right now (and causing issues for users https://github.com/dask/distributed/issues/6228), instead of fixing it, how about we just remove it?

If a worker's connection to the scheduler is broken, have it shut down immediately. This would at least puts us in a consistent state, instead of deadlocking.

Questions:
1. If there's a nanny, should the worker restart?

cc @fjetter @crusaderky @mrocklin

### PR DESCRIPTION
**Pull request by gjoseph92**
When a worker disconnects from the scheduler, close it immediately instead of trying to reconnect.

Also prohibit workers from joining if they have data in memory, as an alternative to https://github.com/dask/distributed/pull/6341.

Just seeing what CI does for now, to figure out which tests to change/remove.

Closes #6350

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by gjoseph92** [comment]
Everything is green except one flaky `test_stress_scatter_death` https://github.com/dask/distributed/issues/6305, and a single `distributed/tests/test_client.py::test_restart_timeout_is_logged` [on Windows](https://github.com/dask/distributed/runs/6496293962?check_suite_focus=true#step:11:1442) is leaking a thread:
```
E               AssertionError: (<Thread(AsyncProcess Dask Worker process (from Nanny) watch process join, started daemon 5744)>, ['  File "C:\\Minico...rocessing\\popen_spawn_win32.py", line 108, in wait
E                 \tres = _winapi.WaitForSingleObject(int(self._handle), msecs)
E                 '])
E               assert False
```
That feels flaky to me, but I guess it's possible this is related?

File: distributed/cli/dask_worker.py [hunk 1]
```diff
@@ -168,11 +168,6 @@
     - 'auto' for automatically computing the memory limit.
     """,
 )
-@click.option(
-    "--reconnect/--no-reconnect",
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/cli/dask_worker.py:172]: I recommend that we err on `--reconnect` and warn on `--no-reconnect`

File: distributed/scheduler.py [hunk 2]
```diff
@@ -3579,6 +3580,16 @@ async def add_worker(
...
 
+        if nbytes:
+            err = (
+                f"Worker {address!r} reconnected with {len(nbytes)} keys in memory! Worker reconnection is not supported. "
+                f"{list(nbytes)}"
+            )
+            logger.error(err)
+            if comm:
+                await comm.write({"status": "error", "message": err, "time": time()})
+            return
```
Comment (mrocklin) [inline distributed/scheduler.py:3591]: Is this likely to occur with workers of this version or later?

File: distributed/scheduler.py [hunk 3]
```diff
@@ -5058,43 +5026,21 @@ async def gather(self, keys, serializers=None):
             )
             result = {"status": "error", "keys": missing_keys}
             with log_errors():
-                # Remove suspicious workers from the scheduler but allow them to
-                # reconnect.
+                # Remove suspicious workers from the scheduler and shut them down.
                 await asyncio.gather(
                     *(
                         self.remove_worker(
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/scheduler.py:5032]: I don't mind be heavy handed, especially if things are in a wonky state.

File: distributed/tests/test_client.py [hunk 4]
```diff
@@ -3589,7 +3589,6 @@ async def test_scatter_raises_if_no_workers(c, s):
         await c.scatter(1, timeout=0.5)
 
 
-@pytest.mark.flaky(reruns=2)
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/tests/test_client.py:3592]: It's because we're selecting a random port.  That port might be used.

File: distributed/tests/test_scheduler.py [hunk 5]
```diff
@@ -1085,101 +1085,12 @@ async def test_worker_arrives_with_processing_data(c, s, a, b):
...
-
-    while len(s.workers) < 3:
-        await asyncio.sleep(0.01)
-    await asyncio.sleep(0.3)
-
-    assert s.get_task_status(keys={x.key, y.key, z.key}) == {
-        x.key: "released",
-        y.key: "memory",
-        z.key: "processing",
-    }
```
Comment (mrocklin) [inline distributed/tests/test_scheduler.py:1180]: Why was this removed?

**Review by mrocklin — COMMENTED** [review]
Incomplete review, but I'm stuck in meetings for the next bit

**Comment by mrocklin** [comment]
In general I'm ok with what's here

File: distributed/tests/test_scheduler.py (hunk 5 above)
Comment (mrocklin) [reply distributed/tests/test_scheduler.py:1180]: Yeah ok, that makes sense.  Thanks.

**Comment by mrocklin** [comment]
> One thing we discussed offline was whether the Nanny should restart the worker in cases when the connection to the scheduler is broken unexpectedly. The answer is probably yes, but as this PR stands, Worker.close will always tell the Nanny it's closing gracefully, preventing restart:

> As a follow-up PR, we should add an option to Worker.close to control this behavior

I may not be understanding the statement above, but we already have the `nanny=` option to `Worker.close`, right?  That's the code snippet that you've highlighted.  I think the question is if there is a specific call of `Worker.close` when we want to pass `nanny=False`.  Agreed?

File: distributed/scheduler.py (hunk 2 above)
Comment (gjoseph92) [reply distributed/scheduler.py:3591]: 807c57aeb1eb8839d01fe998d2ce4d3678a8ade8

**Comment by gjoseph92** [comment]
> we already have the `nanny=` option to `Worker.close`

Great point, I just missed that. Yes, we can set `nanny=False` in these cases here. I'll do that.

**Comment by gjoseph92** [comment]
@mrocklin actually, I think maybe we should do this in a follow-up PR. It's pretty straightforward, but involves a little bit of thinking and a few more tests (the nanny is going to try to report the worker closure to the scheduler at the same time the worker's connection is breaking from it shutting down).

Easy enough, but I'd rather focus on the core change here and not increase the scope.

File: distributed/cli/dask_worker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/cli/dask_worker.py:172]: @mrocklin 3a or 3b?

[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/cli/dask_worker.py:172]: 3a seems like less movement to me.  I would probably default to that.  I'm open to other thoughts though.  I'm honestly not thinking too deeply about this.

[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/cli/dask_worker.py:172]: That's where I'm leaning too, since I don't feel like thinking too deeply about this right now either.

[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/cli/dask_worker.py:172]: Sounds great!

[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/cli/dask_worker.py:172]: I'm going to do this in a separate PR. It's not necessary to get this change in.

[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/cli/dask_worker.py:172]: https://github.com/dask/distributed/pull/6383

**Comment by gjoseph92** [comment]
Failures (both seem flaky):
* https://github.com/dask/distributed/issues/6045 https://github.com/dask/distributed/runs/6516866438?check_suite_focus=true#step:11:1923
* `distributed/cli/tests/test_dask_scheduler.py::test_preload_config ` https://github.com/dask/distributed/runs/6516866324?check_suite_focus=true#step:11:1747

**Comment by gjoseph92** [comment]
@fjetter @mrocklin I believe this is ready for final review. Follow-ups identified in https://github.com/dask/distributed/issues/6384.
